### DISCUSSION TITLE
About Smooth Rotation

### DISCUSSION THREAD

**Discussion by CharlieMalphas** [Techokami/SonicWorldsNext#32] [Q&A] [answered]
Hi, my name is Charlie Malphas and I would like to know how to Smooth Rotate a player character in Sonic Worlds Next. I would appreciate it if you could tell me.

**Comment by DimensionWarped** [comment, marked as answer]
in player.gd
```
# set the sprite to match the sprite rotation variable if it's in the rotatable Sprites list
	if (rotatableSprites.has(animator.current_animation)):
		sprite.rotation = deg2rad(stepify(spriteRotation,45)-90)-rotation-gravityAngle
		# uncomment this next line out for smooth rotation (you should remove the above line too)
		#sprite.rotation = deg2rad(spriteRotation-90)-rotation-gravityAngle
	else:
		sprite.rotation = -rotation+gravityAngle
```

Change this
```
-sprite.rotation = deg2rad(stepify(spriteRotation,45)-90)-rotation-gravityAngle
+sprite.rotation = deg2rad(stepify(spriteRotation,1)-90)-rotation-gravityAngle
```

I think I'll make this a toggle somewhere later. I know a lot of people prefer smooth rotation to the normal stuff.

**Reply by DimensionWarped** [reply]
You could also probably just take the stepify call out entirely if you'd rather btw.

**Reply by CharlieMalphas** [reply]
I was able to do it. Thank you for your answer.

**Reply by UndertaleFan6** [reply]
i tried this and it didn't work
![image](https://github.com/user-attachments/assets/9b99b0ae-f547-4e82-aefa-265b2f3c438d)

**Reply by UndertaleFan6** [reply]
oh i did it wrong
sorry for that

**Comment by Candescence** [comment]
So coming back to this, while the code-based 'toggle' in the current version _does_ work, it's not entirely smooth, since it snaps to the normal of the collision polygon instantly, whereas when people think "smooth" rotation, well, you'd think there'd be a smooth transition between angles.

This is how I have it right now:

```
# set the sprite to match the sprite rotation variable if it's in the rotatable Sprites list
	if (rotatableSprites.has(animator.current_animation)):
		# check if player rotation is greater then 45 degrees or current angle doesn't match the gravity's angle or not on the floor
#		if abs(spriteRotation-90) >= 32 or rotation != gravityAngle or !ground:
#			sprite.rotation = deg_to_rad(snapped(spriteRotation,45)-90)-rotation-gravityAngle
#		else:
#			sprite.rotation = -rotation-gravityAngle
		# uncomment this next line out for smooth rotation (you should remove the above line too)
		sprite.rotation = deg_to_rad(spriteRotation-90)-rotation-gravityAngle
	else:
		sprite.rotation = -rotation+gravityAngle
```

How do I implement fully smoothed rotations? I'm guessing some kind of lerp would be involved.

**Reply by Candescence** [reply]
Okay, so this kinda works:

```
	if (rotatableSprites.has(animator.current_animation)):
		sprite.rotation = sprite.rotation + ((deg_to_rad(spriteRotation-90)-rotation-gravityAngle) - sprite.rotation) * (delta * rotationSpeed)
	else:
		sprite.rotation = sprite.rotation + ((rotation-gravityAngle) - sprite.rotation) * (delta * rotationSpeed)
```

But this particular method doesn't handle transitions to right angle stages properly, you can easily see it on loops. Not sure how to fix that.

**Reply by DimensionWarped** [reply]
I think you're overcomplicating things. If you want Sonic to gradually tilt with the terrain, increase the number of points in your collision mesh for the terrain rather than trying to do some kind of adaptive rotation which is almost certainly going to cause you tons of headaches.
